src/carma-platform/parse_ros_bags.py

```diff
@@ -0,0 +1,94 @@
+import argparse
+import csv
+from pathlib import Path
+
+import more_itertools
+import rosbag
+
+
+def find_msg_closest_to_timestamp(bag_msgs, timestamp):
+    for current, next_ in more_itertools.pairwise(bag_msgs):
+        # Each element in `bag_msgs` is a triple:
+        # <topic_name, msg, system_time_at_receive>
+        _, _, current_log_timestamp = current
+        _, _, next_log_timestamp = next_
+
+        # We only need to iterate until we find two messages who's logged
+        # timestamps straddle the timestamp of interest. All messages
+        # afterwards will have increasing differences.
+        if next_log_timestamp > timestamp:
+            if abs(current_log_timestamp - timestamp) < abs(timestamp - next_log_timestamp):
+                return current
+
+            return next_
+
+
+# calculate the offset between ros clock and the sim clock
+def get_time_offset(ros_bag_file):
+    with rosbag.Bag(ros_bag_file, "r") as bag:
+        cdasim_clock_msgs = bag.read_messages(topics=["/sim_clock"])  # CDASim clock
+        carla_clock_msgs = bag.read_messages(topics=["/clock"])  # CARLA clock
+
+        _, cdasim_clock_msg, cdasim_clock_log_time = next(cdasim_clock_msgs)
+
+        carla_clock_at_cdasim_start = find_msg_closest_to_timestamp(
+            carla_clock_msgs, cdasim_clock_log_time
+        )
+
+        return carla_clock_at_cdasim_start[1].clock - cdasim_clock_msg.clock
+
+
+def get_detected_objects(ros_bag_file, output_file, time_offset):
+    detected_obj_topic_name = "/environment/fused_external_objects"
+    messages = []
+    with rosbag.Bag(ros_bag_file, "r") as bag:
+        for _, msg, _ in bag.read_messages(topics=[detected_obj_topic_name]):
+            messages.extend(msg.objects)
+
+    output_file.parent.mkdir(exist_ok=True)
+
+    if output_file.exists():
+        print(f"Output file {output_file} already exists. Overwriting file.")
+
+    with open(output_file, "w", newline="") as file:
+        writer = csv.writer(file)
+        writer.writerow(
+            ["cdasim_time_ms", "object_id", "map_position_x_m", "map_position_y_m"]
+        )
+        for message in messages:
+            cdasim_time_ms = (message.header.stamp - time_offset).to_sec() * 1_000
+            writer.writerow(
+                [
+                    cdasim_time_ms,
+                    message.id % 1_000,
```

Comment:
This is to get the first 3 digits of the message ID right?

Reply:
This will keep only the last 3 digits. `message.id / 1_000` would remove the last 3 digits.